File: chapt3/topic_practice_ws/src/status_display/src/sys_status_display.cpp
#include <QApplication>  // 导入画界面的工具
#include <QLabel>
#include <QString>
#include <sstream>
#include <thread>
#include "rclcpp/rclcpp.hpp"
#include "status_interfaces/msg/system_status.hpp" // 导入消息格式

using SystemStatus = status_interfaces::msg::SystemStatus;

// 继承自 ROS 节点类，让你拥有收发消息的能力
class SysStatusDisplay : public rclcpp::Node {

private:
    rclcpp::Subscription<SystemStatus>::SharedPtr subscription_;
    QLabel* label_;

public:
    SysStatusDisplay() : Node("sys_status_display") 
    {
        // 步骤 1: 必须先创建 Label，防止回调函数访问空指针
        label_ = new QLabel(get_qstr_from_msg(std::make_shared<SystemStatus>())); 
        label_->setWindowTitle("System Status"); // 给窗口起个名字
        label_->resize(300, 200); // 设置个大小，不然太小了
        label_->show();

        // 步骤 2: 创建订阅者
        subscription_ = this->create_subscription<SystemStatus>(
            "system_status",
            10,
            [&] 
            (const SystemStatus::SharedPtr msg)  -> void { 
            // 注意：这里仍然存在跨线程操作风险，但作为教程练习通常能跑通
            label_->setText(get_qstr_from_msg(msg));
            });

    };

    QString get_qstr_from_msg(const SystemStatus::SharedPtr msg) 
    {
        std::stringstream show_str;
        show_str
            << "======================系统状态可视化显示工具======================" << std::endl
            << "数据时间: \t" << msg->stamp.sec << "\ts\n"
            << "用户名: \t" << msg->host_name << "\t\n"
            << "CPU 使用率: \t" << msg->cpu_percent << "\t%\n"
            << "内存使用率: \t" << msg->memory_percent << "\t%\n"
            << "内存总大小: \t" << msg->memory_total << "\tMB\n"
            << "剩余有效内存: \t" << msg->memory_available << "\tMB\n"
            << "网络发送量: \t" << msg->net_sent << "\tMB\n"
            << "网络接收量: \t" << msg->net_recv << "\tMB\n"
            << "=============================================================" << std::endl;

        return QString::fromStdString(show_str.str());

    };

};


int main(int argc, char *argv[]) 
{
    rclcpp::init(argc, argv);  // 1. 初始化 ROS 2 环境
    QApplication app(argc, argv);  // 2. 初始化 Qt 界面环境
    // 3. 创建你的节点对象 (SysStatusDisplay)
    auto node = std::make_shared<SysStatusDisplay>();
    // 4. 【关键】开启一个新线程去跑 ROS
    // 为什么要新线程？因为下面的 app.exec() 是个死循环。
    // 如果不用新线程，ROS 的 spin 就没机会执行，收不到消息。
    std::thread spin_thread([&]() { rclcpp::spin(node); });
    spin_thread.detach(); // detach 表示“让这个线程自己在后台跑，不管它了”
    app.exec();  // 5. 启动 Qt 的界面循环 (程序会卡在这里，直到你关闭窗口)
    rclcpp::shutdown();   // 6. 退出清理
    return 0;
}

File: chapt3/src/topic_practice_ws/src/status_display_py/status_display_py/sys_status_display.py
import rclpy
import threading
from rclpy.node import Node
from status_interfaces.msg import SystemStatus
from PyQt5.QtWidgets import QApplication, QLabel,QVBoxLayout, QWidget
from PyQt5.QtCore import pyqtSignal, QObject

# 在 ROS 2 + PyQt 的混合开发中始终保持 class MyClass(Node, QWidget): 
# 的顺序，并且在 __init__ 里显式地分别初始化它们

"""
    在 Python 中，当你调用 QWidget.__init__(self) 时，
    QWidget 内部会自动调用 super().__init__() 来寻找下一个父类。 
    根据你定义的顺序(QWidget 在前,Node 在后),
    Python 认为 Node 是 QWidget 的父亲。 
    于是 QWidget 尝试调用 Node 的初始化函数，
    但它没有传任何参数（因为它不知道还要传 node_name),
    导致了 TypeError: missing 1 required positional argument
    所以要注意 Node在前,QWidget在后
"""

class SysStatusDisplay(Node, QWidget):  
    # 定义信号
    update_signal = pyqtSignal(str)
    # 在 Python 中，信号必须定义为类属性（写在 __init__ 外面）
    
    def __init__(self):
        # 初始化父类
        Node.__init__(self, 'sys_status_display')                   
        QWidget.__init__(self)

        # --- 1. 界面初始化  ---
        self.setWindowTitle('系统状态显示')
        self.resize(400, 300)
        
        # 使用布局管理器让界面更规整
        self.layout = QVBoxLayout()
        self.label = QLabel('等待系统状态消息...')
        self.layout.addWidget(self.label)
        self.setLayout(self.layout)
        
        # --- 2. 信号与槽连接 ---
        # 意思：当 update_signal 触发时，自动调用 self.update_label 函数
        self.update_signal.connect(self.update_label)
        
        # --- 3. 创建订阅者 ---
        self.subscription   = self.create_subscription(
            SystemStatus,           # 参数 1: 消息类型 (类名)
            'system_status',        # 参数 2: 话题名
            self.listener_callback,   # 参数 3: 回调函数 (注意这里不加括号！)
            10)                     # 参数 4: QoS (队列长度)
        
     # --- 4. ROS 回调函数 (运行在子线程) ---   
    def listener_callback(self, msg):
        text = f"""
================ 系统状态 (Py) ================
数据时间: \t{msg.stamp.sec} s
主机名:   \t{msg.host_name}
CPU 使用: \t{msg.cpu_percent} %
内存使用: \t{msg.memory_percent} %
内存总计: \t{msg.memory_total:.2f} MB
内存剩余: \t{msg.memory_available:.2f} MB
网络发送: \t{msg.net_sent:.2f} MB
网络接收: \t{msg.net_recv:.2f} MB
=============================================
        """
        
        # 发射信号，把数据传给主线程
        self.update_signal.emit(text)
        
    # --- 5. Qt 槽函数 (运行在主线程) ---
    def update_label(self, text):
        self.label.setText(text)
        
        
def main(args=None):
    # 1. 初始化 ROS 和 Qt
    rclpy.init(args=args)
    app = QApplication([])

    # 2. 创建节点
    node = SysStatusDisplay()
    node.show()  # 窗口显示
    
    # 3. 多线程处理 
    # 必须开一个新线程去跑 ROS 的 spin，否则 Qt 的界面会卡死
    thread = threading.Thread(target=rclpy.spin, args=(node,))
    thread.start()
    
    # 4. 运行 Qt 主循环 
    # 这里的 exec_ 带下划线，是因为 exec 是 Python 关键字
    app.exec_()
    
    # 5. 退出清理
    rclpy.shutdown()
    thread.join()